Source/Core/RequestScheduler.js

```diff
@@ -0,0 +1,383 @@
+    };
+
+    function serverHasOpenSlots(server) {
+        return numberOfActiveRequestsByServer[server] < RequestScheduler.maximumRequestsPerServer;
+    }
+
+    function issueRequest(request) {
+        if (request.state === RequestState.UNISSUED) {
+            request.state = RequestState.ISSUED;
+            request.deferred = when.defer();
+        }
+        return request.deferred.promise;
+    }
+
+    function getRequestReceivedFunction(request) {
+        return function(results) {
+            if (request.state === RequestState.CANCELLED) {
+                return;
+            }
+            --statistics.numberOfActiveRequests;
+            --numberOfActiveRequestsByServer[request.server];
+            request.state = RequestState.RECEIVED;
+            request.deferred.resolve(results);
+        };
+    }
+
+    function getRequestFailedFunction(request) {
+        return function(error) {
+            if (request.state === RequestState.CANCELLED) {
+                return;
+            }
+            ++statistics.numberOfFailedRequests;
+            --statistics.numberOfActiveRequests;
+            --numberOfActiveRequestsByServer[request.server];
+            request.state = RequestState.FAILED;
+            request.deferred.reject(error);
+        };
+    }
+
+    function startRequest(request, passthrough) {
+        if (passthrough) {
+            // Data uri or blob uri should always succeed so make the request immediately and don't contribute to statistics.
+            request.state = RequestState.RECEIVED;
+            return request.requestFunction();
+        }
+
+        var promise = issueRequest(request);
+        request.state = RequestState.ACTIVE;
+        activeRequests.push(request);
+        ++statistics.numberOfActiveRequests;
+        ++statistics.numberOfActiveRequestsEver;
+        ++numberOfActiveRequestsByServer[request.server];
+
+        request.requestFunction().then(getRequestReceivedFunction(request)).otherwise(getRequestFailedFunction(request));
+        return promise;
+    }
+
+    function cancelRequest(request) {
+        if (request.state === RequestState.ACTIVE) {
+            --statistics.numberOfActiveRequests;
+            --numberOfActiveRequestsByServer[request.server];
+            ++statistics.numberOfCancelledActiveRequests;
+            if (defined(request.xhr)) {
+                // TODO : make sure this doesn't trigger a failed promise, if so the deferred can be rejected first
+                // TODO : test this in a test
+                request.xhr.abort();
+            }
+        }
+        request.state = RequestState.CANCELLED;
+        ++statistics.numberOfCancelledRequests;
+        request.deferred.reject('Cancelled');
+    }
+
+    /**
+     * Issuers of a request should update properties of requests. At the end of the frame,
+     * RequestScheduler.update is called to start, cancel, or defer requests.
+     */
+    RequestScheduler.update = function() {
```

Comment:
Since this a static object, but called by the scene.  Is there any concern with calling this once for each frame per-scene instead of once overall?  Would it be better to have `RequestScheduler` use it's own `requestAnimationFrame` callback? This way it gets called once were animation frame for all scenes.